Regression(267014@main) Flaky crash under WebKit::GPUConnectionToWebProcess::didReceiveMessage()

cdumez opened the pull request:
#### e4fcb9c6b04bedb7d92df0b301198c9ab601562b
<pre>
Regression(267014@main) Flaky crash under WebKit::GPUConnectionToWebProcess::didReceiveMessage()
<a href="https://bugs.webkit.org/show_bug.cgi?id=262981">https://bugs.webkit.org/show_bug.cgi?id=262981</a>
rdar://114245301

Reviewed by Simon Fraser.

In 267014@main, we added support for [EnabledIf=Setting] in our IPC messages.in files and
started using it for a few IPC messages. Since then we&apos;ve had flaky crashes on the bots
inside GPUConnectionToWebProcess::didReceiveMessage() because these IPC messages are
sometimes not handled.

The issue is that the checks were racy. The GPUProcess would for example receive the
CreateRemoteGPU IPC and check the `isWebGPUEnabled()` setting. If the setting is disabled,
it would refuse to handle the IPC message and we would crash in debug.

The reason this is racy is that the tests keep enabling / disabling these settings. These
settings are propagated via asynchronous IPC from the UIProcess to the GPUProcess. However,
the IPCs that check for those settings in the GPUProcess are coming from the WebProcess,
not the UIProcess. As a result, it is possible for this to happen:
1. A page gets created with WebGPU enabled
2. The UIProcess sends an async IPC to the GPUProcess to let it know WebGPU is now enabled
3. The WebProcess starts using WebGPU and sends a WebGPU-related IPC to the GPUProcess
4. The GPUProcess receives the WebGPU-related IPC from the WebProcess
5. The GPUProcess receives the IPC from the UIProcess to let it know WebGPU is now enabled

Note that the IPC from the UIProcess (step 5) can get received too late, since there is
no synchronization. The same issue occurs if a page goes away and a setting gets turned
off as a result.

To address the issue, this patch makes it so that the GPUProcessPreferencesForWebProcess
no longer change for a particular WebProcess. This means that if a new page gets created
with different GPUProcess preferences, it will now use a separate WebProcess. Because the
GPUProcess preferences for a particular WebProcess no longer change, there is no race
anymore. When the WebProcess requests a connection to the GPUProcess, it goes via the
UIProcess (synchronously), which sends an IPC to the GPUProcess to create the connection.
The GPUProcessPreferencesForWebProcess are passed with this IPC so that the GPUProcess has
those preferences before receiving any other IPC from the WebProcess.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::allowTestOnlyIPC const):
(WebKit::GPUConnectionToWebProcess::updatePreferences): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updatePreferencesForWebProcess): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updatePreferencesForWebProcess): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::page const):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasSameGPUProcessPreferencesAs const):
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::preferencesForGPUProcess const):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
(WebKit::WebProcessPool::processForRegistrableDomain):
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializePreferencesForGPUProcess):
(WebKit::WebProcessProxy::addProvisionalPageProxy):
(WebKit::WebProcessProxy::addRemotePageProxy):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::computePreferencesForGPUProcess const): Deleted.
(WebKit::WebProcessProxy::updatePreferencesForGPUProcess): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::preferencesForGPUProcess const):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::hasWebPages const):

Canonical link: <a href="https://commits.webkit.org/269209@main">https://commits.webkit.org/269209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b10c69afeff2b419576c94702f92e94387e8bf61

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23771 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20270 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22424 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18983 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24623 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18889 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26109 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20056 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23967 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20502 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19848 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5220 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24052 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
```diff
@@ -52,10 +53,17 @@ static WeakHashSet<SuspendedPageProxy>& allSuspendedPages()
         Ref process = suspendedPage.process();
+#if ENABLE(GPU_PROCESS)
+        auto& preferencesForGPUProcess = process->preferencesForGPUProcess();
+        if (preferencesForGPUProcess && *preferencesForGPUProcess != pageConfiguration.preferencesForGPUProcess())
```
Can this use `hasSameGPUProcessPreferencesAs()` ?

smfr commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1121,10 +1120,19 @@ Ref<WebPageProxy> WebProcessPool::createWebPage(PageClient& pageClient, Ref<API:
 
+    auto haveSameGPUProcessPreferences = [pageConfiguration](const WebPageProxy& relatedPage) {
+#if ENABLE(GPU_PROCESS)
+        return relatedPage.preferencesForGPUProcess() == pageConfiguration->preferencesForGPUProcess();
```
Can this use `hasSameGPUProcessPreferencesAs()`?

In general, there are lot of places in this PR where you compare `preferencesForGPUProcess()`; it would be nice if they all went through some bottleneck function so they are easy to find.

smfr commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1922,8 +1930,19 @@ std::tuple<Ref<WebProcessProxy>, SuspendedPageProxy*, ASCIILiteral> WebProcessPo
+    {
+#if ENABLE(GPU_PROCESS)
+        auto& preferencesForGPUProcess = process.preferencesForGPUProcess();
+        return !preferencesForGPUProcess || page.preferencesForGPUProcess() == *preferencesForGPUProcess;
```
Here's another one.

cdumez committed:
Regression(267014@main) Flaky crash under WebKit::GPUConnectionToWebProcess::didReceiveMessage()
https://bugs.webkit.org/show_bug.cgi?id=262981
rdar://114245301

Reviewed by Simon Fraser.

In 267014@main, we added support for [EnabledIf=Setting] in our IPC messages.in files and
started using it for a few IPC messages. Since then we've had flaky crashes on the bots
inside GPUConnectionToWebProcess::didReceiveMessage() because these IPC messages are
sometimes not handled.

The issue is that the checks were racy. The GPUProcess would for example receive the
CreateRemoteGPU IPC and check the `isWebGPUEnabled()` setting. If the setting is disabled,
it would refuse to handle the IPC message and we would crash in debug.

The reason this is racy is that the tests keep enabling / disabling these settings. These
settings are propagated via asynchronous IPC from the UIProcess to the GPUProcess. However,
the IPCs that check for those settings in the GPUProcess are coming from the WebProcess,
not the UIProcess. As a result, it is possible for this to happen:
1. A page gets created with WebGPU enabled
2. The UIProcess sends an async IPC to the GPUProcess to let it know WebGPU is now enabled
3. The WebProcess starts using WebGPU and sends a WebGPU-related IPC to the GPUProcess
4. The GPUProcess receives the WebGPU-related IPC from the WebProcess
5. The GPUProcess receives the IPC from the UIProcess to let it know WebGPU is now enabled

Note that the IPC from the UIProcess (step 5) can get received too late, since there is
no synchronization. The same issue occurs if a page goes away and a setting gets turned
off as a result.

To address the issue, this patch makes it so that the GPUProcessPreferencesForWebProcess
no longer change for a particular WebProcess. This means that if a new page gets created
with different GPUProcess preferences, it will now use a separate WebProcess. Because the
GPUProcess preferences for a particular WebProcess no longer change, there is no race
anymore. When the WebProcess requests a connection to the GPUProcess, it goes via the
UIProcess (synchronously), which sends an IPC to the GPUProcess to create the connection.
The GPUProcessPreferencesForWebProcess are passed with this IPC so that the GPUProcess has
those preferences before receiving any other IPC from the WebProcess.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
(WebKit::GPUConnectionToWebProcess::allowTestOnlyIPC const):
(WebKit::GPUConnectionToWebProcess::updatePreferences): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updatePreferencesForWebProcess): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/Shared/GPUProcessConnectionParameters.h:
* Source/WebKit/Shared/GPUProcessConnectionParameters.serialization.in:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.h:
* Source/WebKit/Shared/GPUProcessPreferencesForWebProcess.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::preferencesForGPUProcess const):
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::updatePreferencesForWebProcess): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.h:
(WebKit::RemotePageProxy::page const):
* Source/WebKit/UIProcess/SuspendedPageProxy.cpp:
(WebKit::SuspendedPageProxy::findReusableSuspendedPageProcess):
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::hasSameGPUProcessPreferencesAs const):
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::preferencesForGPUProcess const):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessCache.cpp:
(WebKit::WebProcessCache::takeProcess):
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createGPUProcessConnection):
(WebKit::WebProcessPool::processForRegistrableDomain):
(WebKit::WebProcessPool::createWebPage):
(WebKit::WebProcessPool::processForNavigationInternal):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::initializePreferencesForGPUProcess):
(WebKit::WebProcessProxy::addProvisionalPageProxy):
(WebKit::WebProcessProxy::addRemotePageProxy):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::createGPUProcessConnection):
(WebKit::WebProcessProxy::computePreferencesForGPUProcess const): Deleted.
(WebKit::WebProcessProxy::updatePreferencesForGPUProcess): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::preferencesForGPUProcess const):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::create):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::hasWebPages const):

Canonical link: https://commits.webkit.org/269209@main
